### PR TITLE
Fix for GHC 9.2

### LINKED ISSUE #311: Compatibility with GHC 9.2
Could you upload a compatible release?

With `--allow-newer`, I currently get an error:

```
src/Snap/Internal/Parsing.hs:439:64: error:
    • Couldn't match expected type ‘ghc-prim-0.8.0:GHC.Prim.Word#’
                  with actual type ‘ghc-prim-0.8.0:GHC.Prim.Word8#’
    • In the first argument of ‘uncheckedShiftRL#’, namely ‘a#’
      In the first argument of ‘word2Int#’, namely
        ‘(uncheckedShiftRL# a# b#)’
      In the first argument of ‘I#’, namely
        ‘(word2Int# (uncheckedShiftRL# a# b#))’
    |
439 |     shiftr (W8# a#) (I# b#) = I# (word2Int# (uncheckedShiftRL# a# b#))
    |                                                                ^^
cabal: Failed to build snap-core-1.0.4.2
```

### PR DESCRIPTION
**Pull request by brandon-leapyear**
Resolves #311 

Supercedes #304

### CODE REVIEW

**Comment by danwdart** [comment]
Ah, and:
```
running tests
Running 1 test suites...
Test suite testsuite: RUNNING...
Snap.Internal.Http.Types.Tests:
  httpTypes/show: [OK]
  httpTypes/cookies: [OK]
  httpTypes/cookieToBS: [OK]
  httpTypes/urlDecode: [OK]
  httpTypes/formatLogTime: [OK]
  httpTypes/addHeader: [OK]
  httpTypes/methodOrd: [OK]
Snap.Internal.Routing.Tests:
  route/1: [OK]
  route/2: [OK]
  route/3: [OK]
  route/4: [OK]
  route/5: [OK]
  route/6: [OK]
  route/7: [OK]
  route/8: [OK]
  route/9: [OK]
  route/10: [OK]
  route/11: [OK]
  route/12: [OK]
  route/13: [OK]
  route/14: [OK]
  route/15: [OK]
  route/16: [OK]
  route/17: [OK]
  route/18: [OK]
  route/19: [OK]
  route/20: [OK]
  route/21: [OK]
  route/22: [OK]
  route/23: [OK]
  route/24: [OK]
  route/25: [OK]
  route/26: [OK]
  route/27: [OK]
  route/28: [OK]
  route/routeLocal: [OK]
  route/urlDecode: [OK]
  route/urlEncodedPath: [OK]
  route/emptyCapture: [OK]
  route/captureConflicts: [OK]
  route/trivials: [OK]
  route/paramUnification: [OK]
  route/failedUrlDecode: [OK]
  route/dirFallthrough: [OK]
Snap.Core.Tests:
  failure: [OK]
  core/alternative: [OK]
  core/earlyTermination: [OK]
  core/escapeHttp: [OK]
  core/catchFinishWith: [OK]
  core/requestBodies: [OK]
  core/requestBodyException: [OK]
  core/requestBodyTermination: [OK]
  core/requestBodyTooLong: [OK]
  core/requestBodyTooSlow: [OK]
  core/trivials: [OK]
  core/method: [OK]
  core/methods: [OK]
  core/Method/eq: [OK, passed 100 tests]
  core/Method/noteq: [OK, passed 100 tests]
  core/dir: [OK]
  core/catchIO: [OK]
  core/writes: [OK]
  core/getParam: [OK]
  core/urlEncoding1: [OK]
  core/urlEncoding2: [OK, passed 100 tests]
  core/dir2: [OK]
  core/ipHeaderFilter: [OK]
  core/mzero404: [OK]
  core/evalSnap-exception: [OK]
  core/localRequest: [OK]
  core/redirect: [OK]
  core/bracketSnap: [OK]
  core/instances: [OK]
  core/pathArgs: [OK]
  core/stateT_exceptions: [OK]
  core/304fixup: [OK]
  core/chunked-fixup: [OK]
Snap.Internal.Parsing.Tests:
  parsing/avpairs: [OK]
  parsing/parseCookie: [OK]
  parsing/headers: [OK]
  parsing/quoted-string: [OK]
  parsing/quoted-string-utf8: [OK]
  parsing/unsafeFromHex: [OK]
  parsing/unsafeFromNat: [OK]
  parsing/urlEncoded: [OK]
  parsing/failParse: [OK]
  parsing/tokens: [OK]
Snap.Types.Headers.Tests:
  types/headers/show: [OK]
  types/headers/toFrom: [OK]
  types/headers/folds: [OK]
Snap.Util.FileServe.Tests:
  fileServe/foo.bin: [Failed]
fileServe/foo.bin/ifModifiedSince/contents
expected: "FOO\n"
 but got: ""
  fileServe/foo.txt: [OK]
  fileServe/foo.html: [OK]
  fileServe/foo.bin.bin.bin: [OK]
  fileServe/directory-passes: [OK]
  fileServe/404s: [OK]
  fileServe/Single: [OK]
  fileServe/cfgA: [OK]
  fileServe/cfgB: [OK]
  fileServe/cfgC: [OK]
  fileServe/cfgD: [OK]
  fileServe/cfgFancy: [OK]
  fileServe/range/ok: [OK]
  fileServe/range/bad: [OK]
  fileServe/range/multi: [OK]
  fileServe/range/if-range: [Failed]
foo.bin 200
expected: 200
 but got: 206
  fileServe/badUrl: [OK]
  fileServe/decodeFilePath: [OK]
Snap.Util.FileUploads.Tests:
  fileUploads/fold1: [OK]
  fileUploads/fold2: [OK]
  fileUploads/filePolicyViolation1: [OK]
  fileUploads/filePolicyViolation2: [OK]
  fileUploads/emptyNamePolicyViolation: [OK]
  fileUploads/emptyNameStore: [OK]
  fileUploads/emptyNameSkip: [OK]
  fileUploads/temporaryStore: [OK]
  fileUploads/temporaryStoreSafeMove: [OK]
  fileUploads/success1: [OK]
  fileUploads/success2: [OK]
  fileUploads/utf8-filename: [OK]
  fileUploads/badParses: [OK]
  fileUploads/perPartPolicyViolation1: [OK]
  fileUploads/perPartPolicyViolation2: [OK]
  fileUploads/formInputsPolicy: [OK]
  fileUploads/formSizePolicy: [OK]
  fileUploads/noFileName: [OK]
  fileUploads/noFileNameTooBig: [OK]
  fileUploads/tooManyHeaders: [OK]
  fileUploads/noBoundary: [OK]
  fileUploads/noMixedBoundary: [OK]
  fileUploads/wrongContentType: [OK]
  fileUploads/tooSlow: [OK]
  fileUploads/tooSlow2: [OK]
  fileUploads/abortedBody: [OK]
  fileUploads/trivials: [OK]
  fileUploads/disconnectionCleanup: [OK]
Snap.Util.GZip.Tests:
  gzip/identity1: [OK, passed 100 tests]
  gzip/identity1_charset: [OK, passed 100 tests]
  gzip/identity2: [OK, passed 100 tests]
  gzip/identity3: [OK, passed 100 tests]
  gzip/identity4: [OK, passed 100 tests]
  gzip/identity5: [OK, passed 100 tests]
  gzip/noheaders: [OK, passed 100 tests]
  gzip/noAcceptEncoding: [OK, passed 100 tests]
  gzip/acceptEncodingBad: [OK]
  gzip/testNopWhenContentEncodingSet: [OK, passed 100 tests]
  gzip/testCompositionDoesn'tExplode: [OK, passed 100 tests]
  gzip/lotsOfChunks: [OK]
  gzip/noCompression: [OK, passed 100 tests]
  gzip/bad headers: [OK, passed 100 tests]
  gzip/trivials: [OK]
Snap.Util.Proxy.Tests:
  proxy/no-proxy: [OK]
  proxy/forwarded-for: [OK]
  proxy/trivials: [OK]
Snap.Util.CORS.Tests:
  CORS/simple: [OK]
  CORS/options: [Failed]
Header "Access-Control-Allow-Headers"
expected: Just "Content-Type, X-STUFF"
 but got: Just "X-STUFF, Content-Type"
Snap.Test.Tests:
  test/requestBuilder/defaultBuild: [OK]
  test/requestBuilder/reqToString: [OK]
  test/requestBuilder/setRequestType: [OK]
  test/requestBuilder/testSetQueryString: [OK]
  test/requestBuilder/testSetQueryStringRaw: [OK]
  test/requestBuilder/testHeaders: [OK]
  test/requestBuilder/testMisc: [OK]
  test/requestBuilder/testMultipart: [OK]
  test/requestBuilder/testPost: [OK]
  test/requestBuilder/postFormEncoded: [OK]
  test/requestBuilder/testToString: [OK]
  test/requestBuilder/testAssert404: [OK]
  test/requestBuilder/testAssertBodyContains: [OK]
  test/requestBuilder/testAssertRedirect: [OK]
  test/requestBuilder/cookies: [OK]
  test/requestBuilder/terminate: [OK]
  test/requestBuilder/trivials: [OK]
  test/requestBuilder/hostName: [OK]

         Properties   Test Cases    Total
 Passed  15           156           171
 Failed  0            3             3
 Total   15           159           174

Test suite testsuite: FAIL
```

but don't know if it's going to have been related.

**Comment by brandon-leapyear** [comment]
I got it working with the old version of hashable; the new version of hashable is not strictly necessary for GHC 9.2. The upper bound for hashable can be bumped in a separate PR.

The `CORS/options` test failure seems to also fail on master. Seems like the test shouldn't depend on the order of those values, since it's built from a `HashSet` and order isn't guaranteed. I can fix the test here. I can't seem to repro the other failures you got (e.g. the `fileServe` tests)

Also, I built locally with Stack and had to make a `stack.yaml` file to do so. Do you want me to commit that file?

**Comment by danwdart** [comment]
Thanks, I don't need stack.yaml files, but if they're to be supported, it sounds like good practice.

**Comment by ysangkok** [comment]
@danwdart What needs to be done for this to get merged?

File: snap-core.cabal [hunk 1]
```diff
@@ -148,7 +148,7 @@ Library
     readable                  >= 0.1     && < 0.4,
     regex-posix               >= 0.95    && < 1,
     text                      >= 0.11    && < 1.3,
-    time                      >= 1.0     && < 1.11,
```
Comment (ysangkok) [inline snap-core.cabal:151]: May as well support time 1.12 too since it is already released.

**Review by ysangkok — CHANGES_REQUESTED** [review]

File: snap-core.cabal [hunk 2]
```diff
@@ -148,7 +148,7 @@ Library
...
+    time                      >= 1.0     && < 1.12,
```
Comment (ysangkok) [inline snap-core.cabal:151]: May as well support time 1.12 too since it is already released.

**Review by ysangkok — CHANGES_REQUESTED** [review]

**Comment by ysangkok** [comment]
Interesting how the header order was also swapped in the recent PR #307 by @kozross . I don't know where the non-determinism is coming from but surely it must be ok to not have the test suite care, as you have suggested.

**Comment by brandon-leapyear** [comment]
Yes, I'm not sure why the headers are swapping. I just pushed up a change to reduce the diff

**Comment by ysangkok** [comment]
@gregorycollins Any chance you could take a look at this, or do you know who I can ping? Thanks.

**Pull request merged**